fix: pass through NaN values when building graph from weighted adjacency matrix

szhorvat opened the pull request:
Ref #2413 

NaN values are now passed through when building a _weighted_ graph  from an adjacency matrix.

This still needs tests, but I'd like a review now @ntamas.

There were two issues that needed to be addressed:

 - max(NaN, anything) should give NaN with the "max" method
 - Adjacency matrices with NaN in symmetric positions should be treated as symmetric. `igraph_matrix_is_symmetric()` doesn't give `true` for these because `NaN != NaN`. A clean way to change the behaviour of `igraph_matrix_is_symmetric()` would be to [define the `EQ` macro](https://github.com/igraph/igraph/blob/master/src/core/matrix.pmt#L1579) so that it treats NaN as equal to NaN, but this would have further implications which are undesirable. In almost all contexts, we want NaN != NaN. Therefore I re-implemented the symmetry check.

It is interesting to note that some systems, like MATLAB, consider any matrix with NaNs to be non-symmetric. R, however, does accept them as symmetric.

```
> isSymmetric(m)
[1] TRUE
> m<-cbind(c(0,NaN),c(NaN,0))
> isSymmetric(m)
[1] TRUE
> NA==NA
[1] NA
> NaN==NaN
[1] NA
```

It seems safer to me to not change `igraph_matrix_is_symmetric()` in igraph and let it continue rejecting NaNs. This is actually useful in numerical contexts.

ntamas commented on src/constructors/adjacency.c:
```diff
@@ -531,11 +531,21 @@ static igraph_error_t igraph_i_weighted_adjacency_undirected(
+     * returns false for these as NaN != NaN. */
+    igraph_integer_t n = igraph_matrix_nrow(adjmatrix);
+    for (igraph_integer_t i=0; i < n; i++) {
+        for (igraph_integer_t j=0; j < n; j++) {
```
Aren't we comparing each `(i, j)` pair twice this way? I think it is enough to start `j` from `i+1`, right?

szhorvat commented:
Rebased on current `master`.

szhorvat commented:
@ntamas I don't want to leave this PR hanging forever. What we have should fix NaN passthrough in `igraph_weighted_adjacency()`. Let's merge this and deal with the rest later. I added TODO items in #2413 and added tests to for the fixes in this PR.
